torch/distributed/pipelining/schedules.py

```diff
@@ -1638,21 +1658,38 @@ def _assert_unsharded(stage_idx: int):
                     if stage_uses_fsdp:
                         _assert_unsharded(stage_idx)
 
-                    if not stage.is_first:
+                    if (
+                        not stage.is_first
+                        # no recv op expected for V-schedule special case (see [Note: V-schedule special case])
+                        and stage_idx - 1 not in stage_index_to_stage
```

Comment:
nit: for better clarity the check `stage_idx +/- 1 in stage_index_to_stage` which is recurring in a few spots could be refactored into a property like `has_local_next_stage` and `has_local_prev_stage`